designs/delay-node-enhancements/delay-node-enhancements.md

```diff
@@ -0,0 +1,83 @@
+---
+state: draft
+---
+
+# delay node enhancements
+## Summary
+ The enhancements for delay node change message queue handling and provide logging function.
+
+## Author
+ @kazuhitoyokoi
+
+## Details
+### Use cases
+#### 1. Incoming message control
+ Because Node-RED accepts all incoming message from input nodes like mqtt-in node, Node.js process sometimes reaches the upper limit of memory usage.
+To avoid the risk of message queue overflow, rate limit functionality in delay node will be useful but Node-RED users cannot customize queue handling.
+For example, the current delay node cannot simultaneously use the drop option and queue to process burst incoming messages.
+
+#### 2. Changing properties on demand
+ Currently, the delay node supports msg.delay to change property value on demand.
+If the delay node additionally supports msg.rate to set rate value and msg.queueLength to change queue size, users can create custom delay node using subflow to wrap delay node.
+
+#### 3. Logging and error handling
+ Because current drop option in rate limit mode doesn't record dropped messages, Node-RED users aren't aware of the lost messages.
+Therefore, it will be useful for Node-RED users to track background behaviors if the delay node outputs the number of dropped messages to the log stream.
+Additionally, Node-RED users can add their error handling in their flow if the delay node throws the dropped messages to the catch node when the queue reaches the defined limit.
+
+### Options in setting.js which the delay node uses
+- nodeMessageBufferMaxLength
+
+  The enhanced delay node supports nodeMessageBufferMaxLength as the default buffer length.
+  The buffer length needs to be greater than queue length defined in msg.queueLength or node property UI.
+
+- customQueueLength (default: false)
```

Comment:
Again this is now global - so again if you had a subflow it would be enable for all or none.

Reply:
I changed the `customQueueLength` to `nodeMessageBufferMaxLength` in the settings.js.